src/appleseed/foundation/array/array.cpp

```diff
@@ -0,0 +1,209 @@
+#include "foundation/utility/otherwise.h"
+
+// Stamdard headers.
+#include <algorithm>
+
+using namespace std;
+
+namespace foundation
+{
+
+//
+// Array class implementation.
+//
+
+Array::Concept::~Concept()
+{
+}
+
+Array::Array(const ArrayType type, const size_t n)
+{
+#define ARRAY_INIT_CASE(type_enum, type_name)\
+    case type_enum:\
+        m_self = new Model<type_name>();\
+    break;
+
+    switch (type)
+    {
+      ARRAY_INIT_CASE(UInt8Type, uint8)
+      ARRAY_INIT_CASE(UInt16Type, uint16)
+      ARRAY_INIT_CASE(UInt32Type, uint32)
+      ARRAY_INIT_CASE(FloatType, float)
+      ARRAY_INIT_CASE(Vector2fType, Vector2f)
+      ARRAY_INIT_CASE(Vector3fType, Vector3f)
+      ARRAY_INIT_CASE(CompressedUnitVectorType, CompressedUnitVector)
+      ARRAY_INIT_CASE(Color3fType, Color3f)
+
+      assert_otherwise;
+    }
+
+#undef ARRAY_INIT_CASE
+
+    if (n != 0)
+        resize(n);
+}
+
+Array::~Array()
+{
+    delete m_self;
+}
+
+Array::Array(const Array& other)
+{
+    m_self = other.m_self->copy();
+}
+
+Array::Array(Array&& other) APPLESEED_NOEXCEPT
+{
+    m_self = other.m_self;
+    other.m_self = nullptr;
+}
+
+Array& Array::operator=(const Array& other)
+{
+    Concept* tmp(other.m_self->copy());
+
+    delete m_self;
+    m_self = tmp;
+
+    return *this;
+}
+
+Array& Array::operator=(Array&& other) APPLESEED_NOEXCEPT
+{
+    swap(m_self, other.m_self);
+    return *this;
+}
+
+ArrayType Array::type() const
+{
+    assert(!is_moved());
+    return m_self->type();
+}
+
+bool Array::empty() const
+{
+    assert(!is_moved());
+    return m_self->empty();
+}
+
+size_t Array::size() const
+{
+    assert(!is_moved());
+    return m_self->size();
+}
+
+size_t Array::capacity() const
+{
+    assert(!is_moved());
+    return m_self->capacity();
+}
+
+void Array::clear()
+{
+    assert(!is_moved());
+    m_self->clear();
+}
+
+void Array::resize(const size_t n)
+{
+    assert(!is_moved());
+    m_self->resize(n);
+}
+
+void Array::reserve(const size_t n)
+{
+    assert(!is_moved());
+    m_self->reserve(n);
+}
+
+void Array::shrink_to_fit()
+{
+    assert(!is_moved());
+    m_self->shrink_to_fit();
+}
+
+void Array::push_back(const void* p)
+{
+    assert(!is_moved());
+    m_self->push_back(p);
+}
+
+const void* Array::begin() const
+{
+    assert(!is_moved());
+    return m_self->begin();
+}
+
+const void* Array::end() const
+{
+    assert(!is_moved());
+    return m_self->end();
+}
+
+void* Array::begin()
+{
+    assert(!is_moved());
+    return m_self->begin();
+}
+
+void* Array::end()
+{
+    assert(!is_moved());
+    return m_self->end();
+}
+
+bool Array::operator==(const Array& rhs) const
+{
+    assert(!is_moved());
+    assert(!rhs.is_moved());
+
+    return type() == rhs.type() && m_self->equals(rhs.m_self);
+}
+
+bool Array::operator!=(const Array& rhs) const
+{
+    assert(!is_moved());
+    assert(!rhs.is_moved());
+
+    return type() != rhs.type() || !m_self->equals(rhs.m_self);
+}
+
+bool Array::is_moved() const
```

Comment:
The handling of moved arrays seems a bit peculiar, can you explain the logic? Why do we need this method?

Reply:
What do you mean by peculiar? I don't think we are doing anything strange in the move constructor or assigment operator.

We assert this all the time, everywhere in the Array class and we use it in the array tests.